src/mono/mono/metadata/icall.c

```diff
@@ -233,6 +233,31 @@ ves_icall_System_Array_SetValueRelaxedImpl (MonoArrayHandle arr, MonoObjectHandl
 	array_set_value_impl (arr, value, pos, FALSE, FALSE, error);
 }
 
+void
+ves_icall_System_Array_InitializeInternal (MonoObjectHandleOnStack arr_handle, MonoError *error)
```

Comment:
@reflectronic Are you up for implementing the mono version mostly in C#, too?

I think we can add something like NativeAOT's default constructor -> function pointer icall and then most of this code could be managed.

We don't need to do it in this PR (unless you want to). Just an idea.

Reply:
I would have done it like that, but I couldn't quite find the function I wanted (wasn't sure if the thing I was looking at was safe to just call with `calli`, whether it would try to unbox the argument, etc.). Certainly, I could address it in a follow up.

Reply:
@reflectronic I don't think there's an existing icall, but it should be ok to add something like this:

```
MonoMethod * const method = mono_class_get_method_from_name_checked (element_class, ".ctor", 0, 0, error);
if (!method) {
	return NULL;
}
gpointer addr = mono_compile_method_checked (method, error);
if (!is_ok (error)) {
	mono_error_cleanup (error);
	return NULL;
}
return mono_create_ftnptr(addr);
```

on the C# side it would look pretty much like the NativeAOT version - cast the icall's `IntPtr` result to a function pointer type and call it.

> wasn't sure if the thing I was looking at was safe to just call with calli, whether it would try to unbox the argument, etc.

I'm pretty certain that it will just work. I believe it should be expecting an unboxed argument.